pattern/equivalence/AlphaEquivalence.java

```diff
@@ -17,262 +17,115 @@
+        assert from.reference().isName() == to.reference().isName();
+        Map<Variable, Variable> newMap = variableMapping();
+        newMap.put(from, to);
+        Map<Variable, Variable> reverseMap = reverseVariableMapping();
+        reverseMap.put(to, from);
+        return new AlphaEquivalence(newMap, reverseMap);
     }
 
-    public static class EquivalenceSet<T extends AlphaEquivalent<T>> implements AlphaEquivalent<EquivalenceSet<T>> {
-
-        private final Set<T> set;
-
-        private EquivalenceSet(Set<T> set) {
-            this.set = set;
-        }
-
-        public static <S extends AlphaEquivalent<S>> EquivalenceSet<S> of(Set<S> set) {
-            return new EquivalenceSet<>(set);
-        }
-
-        @Override
-        public AlphaEquivalence alphaEquals(EquivalenceSet<T> that) { // TODO: Should be able to accept a set not an alpha set?
-            if (that.size() != size()) return AlphaEquivalence.invalid();
-            try {
-                return containsAll(that);
-            } catch (NullPointerException unused) {
-                return AlphaEquivalence.invalid();
-            }
-        }
-
-        private int size() {
-            return set.size();
-        }
+    public static  <T extends AlphaEquivalent<T>> FunctionalIterator<AlphaEquivalence> alphaEquals(@Nullable T member1,
+                                                                                                   @Nullable T member2) {
+        if (member1 == null && member2 == null) return Iterators.single(empty());
+        if (member1 == null || member2 == null) return Iterators.empty();
+        else return member1.alphaEquals(member2);
+    }
 
-        private Iterator<T> iterator() {
-            return set.iterator();
-        }
+    public Map<Variable, Variable> variableMapping() {
+        return new HashMap<>(map);
+    }
 
-        private AlphaEquivalence containsAll(EquivalenceSet<T> c) {
-            AlphaEquivalence alphaMap = Valid.create();
-            for (T e : c.set()) alphaMap = alphaMap.addOrInvalidate(contains(e));
-            if (!alphaMap.isValid()) return alphaMap;
-            return alphaMap;
-        }
+    private Map<Variable, Variable> reverseVariableMapping() {
+        return new HashMap<>(reverseMap);
+    }
 
-        private AlphaEquivalence contains(T o) {
-            Iterator<T> it = iterator();
-            if (o == null) {
-                while (it.hasNext()) {
-                    if (it.next() == null) return Valid.create();
-                }
-            } else {
-                while (it.hasNext()) {
-                    AlphaEquivalence alphaMap = it.next().alphaEquals(o);
-                    if (alphaMap.isValid()) return Valid.create(alphaMap.asValid().variableMapping());
-                }
-            }
-            return AlphaEquivalence.invalid();
-        }
+    public Map<Retrievable, Retrievable> retrievableMapping() {
```

Comment:
If this class can only contain Variables with named identifiers, the mapping of identifiers returned here can be tighter: there's a subtype called `Identifier.Retrievable.Name`

Reply:
This isn't the case! When we were reviewing I think you said you picked this up from `assert from.reference().isName() == to.reference().isName();` but this assertion says that both variables must either both be named or both anonymous. In general for alpha equivalence the mappings can contain a mixture of both named and anonymous variables.